nixos/modules/tasks/filesystems/btrfs.nix

```diff
@@ -1,35 +1,130 @@
 
-    system.fsPackages = [ pkgs.btrfs-progs ];
+      fileSystems = mkOption {
+        type = types.listOf types.path;
+        example = [ "/" ];
+        description = ''
+          List of paths to btrfs filesystems to regularily call <command>btrfs scrub</command> on.
+          Defaults to all mount points with btrfs filesystems.
+          If you mount a filesystem multiple times or additionally mount subvolumes,
+          you need to manually specify this list to avoid scrubbing multiple times.
+        '';
+      };
 
-    boot.initrd.kernelModules = mkIf inInitrd [ "btrfs" "crc32c" ];
+      interval = mkOption {
+        default = "monthly";
+        type = types.str;
+        example = "weekly";
+        description = ''
+          Systemd calendar expression for when to scrub btrfs filesystems.
+          The recommended period is a month but could be less
+          (<citerefentry><refentrytitle>btrfs-scrub</refentrytitle>
+          <manvolnum>8</manvolnum></citerefentry>).
+          See
+          <citerefentry><refentrytitle>systemd.time</refentrytitle>
+          <manvolnum>7</manvolnum></citerefentry>
+          for more information on the syntax.
+        '';
+      };
+
+    };
+  };
 
-    boot.initrd.extraUtilsCommands = mkIf inInitrd
+  config = mkMerge [
+    (mkIf enableBtrfs {
+      system.fsPackages = [ pkgs.btrfs-progs ];
+
+      boot.initrd.kernelModules = mkIf inInitrd [ "btrfs" "crc32c" ];
+
+      boot.initrd.extraUtilsCommands = mkIf inInitrd
       ''
         copy_bin_and_libs ${pkgs.btrfs-progs}/bin/btrfs
         ln -sv btrfs $out/bin/btrfsck
         ln -sv btrfsck $out/bin/fsck.btrfs
       '';
 
-    boot.initrd.extraUtilsCommandsTest = mkIf inInitrd
+      boot.initrd.extraUtilsCommandsTest = mkIf inInitrd
       ''
         $out/bin/btrfs --version
       '';
 
-    boot.initrd.postDeviceCommands = mkIf inInitrd
+      boot.initrd.postDeviceCommands = mkIf inInitrd
       ''
         btrfs device scan
       '';
-  };
+    })
+
+    (mkIf enableAutoScrub {
+      assertions = [
+        {
+          assertion = cfgScrub.enable -> (cfgScrub.fileSystems != []);
+          message = ''
+            If 'services.btrfs.autoScrub' is enabled, you need to have at least one
+            btrfs file system mounted via 'fileSystems' or specify a list manually
+            in 'services.btrfs.autoScrub.fileSystems'.
+          '';
+        }
+      ];
+
+      # This will yield duplicated units if the user mounts a filesystem multiple times
+      # or additionally mounts subvolumes, but going the other way around via devices would
+      # yield duplicated units when a filesystem spans multiple devices.
+      # This way around seems like the more sensible default.
+      services.btrfs.autoScrub.fileSystems = mkDefault (mapAttrsToList (name: fs: fs.mountPoint)
+      (filterAttrs (name: fs: fs.fsType == "btrfs") config.fileSystems));
+
+      # did not manage to do it via the usual btrfs-scrub@.timer/.service
```

Comment:
I'm curious, what issues did you encounter using the template units?

Reply:
(Not a blocker, just to be clear)

Reply:
@joachifm the full story:

What I initially tried to achieve was this, which seems to be the usual solution on other distros (copied from Arch Linux):

`btrfs-scrub@.timer`:
```
[Unit]
Description=Monthly Btrfs scrub on %f

[Timer]
OnCalendar=monthly
AccuracySec=1d
Persistent=true

[Install]
WantedBy=multi-user.target
```

`btrfs-scrub@.service`:
```
[Unit]
Description=Btrfs scrub on %f

[Service]
Nice=19
IOSchedulingClass=idle
ExecStart=/usr/bin/btrfs scrub start -B %f
```

enabling specific units for scrubbing:
```
> systemctl enable btrfs-scrub@$(systemd-escape /).timer
Created symlink /etc/systemd/system/multi-user.target.wants/btrfs-scrub@-.timer → /usr/lib/systemd/system/btrfs-scrub@.timer.
> systemctl enable btrfs-scrub@$(systemd-escape /home).timer
Created symlink /etc/systemd/system/multi-user.target.wants/btrfs-scrub@-home.timer → /usr/lib/systemd/system/btrfs-scrub@.timer.
```
each parameterized timer unit will execute a corresponding parameterized service unit.

My issues:

Packing the units (without [Install], of course) works fine using `systemd.services/timers."btrfs-scrub@"`. But enabling the parameterized timers was problematic: I came up with a few variants (should have kept the code to show it to you…), trying to use `wants` / `wantedBy` and `aliases` to create the required symlinks. But as soon as something like that is added to the timer, the timer file suddenly is empty like this:

```
[Unit]

[Timer]
```
without any error message. The only thing I noted in the logs was that sometimes it tried to start `btrfs-scrub@timers.timer`, but I have no idea where the `timers` parameterization came from - from `timers.target`? But how?

(I found no exapmle in NixOS where something similar is done.)

So I tried to have only a single timer, activating a list of services, but a timer can only start a single services.  Which lead me to the non-templated version you see here.

Reply:
Hm, alright, well I guess this can be rectified later.

Reply:
Okay, I was not sure whether that is actually wanted or preffered vs letting nix do all the work.
I extended the comment a little describing the issue, and marked it as a TODO.